frontend/controller/actions/identity.js

```diff
@@ -469,7 +428,51 @@ export default (sbp('sbp/selectors/register', {
         }).catch((e) => {
           console.error('Error at gi.ui/prompt', e)
         })
-      })
+      }).then(() =>
+        sbp('chelonia/contract/sync', contractIDs).then(async function () {
+          // contract sync might've triggered an async call to /remove, so wait before proceeding
+          await sbp('chelonia/contract/wait', contractIDs)
+          // similarly, since removeMember may have triggered saveOurLoginState asynchronously,
+          // we must re-sync our identity contract again to ensure we don't rejoin a group we
+          // were just kicked out of
+          await sbp('chelonia/contract/sync', identityContractID, { force: true })
+          await sbp('gi.actions/identity/updateLoginStateUponLogin')
+          await sbp('gi.actions/identity/saveOurLoginState') // will only update it if it's different
+
+          // The state above might be null, so we re-grab it
+          const state = sbp('state/vuex/state')
+
+          // Call 'gi.actions/group/join' on all groups which may need re-joining
+          await Promise.all(groupsToRejoin.map(groupId => {
+            return (
+            // (1) Check whether the contract exists (may have been removed
+            //     after sync)
+              state.contracts[groupId] &&
+            // (2) Check whether the join process is still incomplete
+            //     This needs to be re-checked because it may have changed after
+            //     sync
+            !state.profiles?.[username] &&
+            // (3) Call join
+            sbp('gi.actions/group/join', { contractID: groupId, contractName: 'gi.contracts/group' })
+            )
+          }))
+
+          // update the 'lastLoggedIn' field in user's group profiles
+          sbp('state/vuex/getters').groupsByName
+            .map(entry => entry.contractID)
+            .forEach(cId => {
+            // We send this action only for groups we have fully joined (i.e.,
+            // accepted an invite add added our profile)
+              if (state[cId]?.profiles?.[username]) {
+                sbp('gi.actions/group/updateLastLoggedIn', { contractID: cId }).catch(console.error)
+              }
+            })
+        }).catch((err) => {
+          console.error('Error during contract sync upon login', err)
+        }).finally(() => {
+          sbp('okTurtles.events/emit', LOGIN, { username, identityContractID })
+        })
```

Comment:
This is different from what was there before, and therefore changes the behavior of the UI from how it worked before, possibly breaking the tests because it claims we're logged in when we're not.